smarts/core/smarts.py

```diff
@@ -108,16 +109,28 @@ def __init__(
         self._motion_planner_provider = MotionPlannerProvider()
         self._traffic_history_provider = TrafficHistoryProvider()
         self._trajectory_interpolation_provider = TrajectoryInterpolationProvider()
-        self._providers = [
+        self._provider_recovery_flags: Dict[Provider, ProviderRecoveryFlags] = {}
+        self._providers: List[Provider] = []
+
+        self.add_provider(
             self._motion_planner_provider,
+        )
+        self.add_provider(
             self._traffic_history_provider,
+        )
+        self.add_provider(
             self._trajectory_interpolation_provider,
-        ]
+        )
         if self._traffic_sim:
-            self._providers.insert(0, self._traffic_sim)
+            self._insert_provider(
+                0,
+                self._traffic_sim,
+                recovery_flags=ProviderRecoveryFlags.EPISODE_REQUIRED
+                | ProviderRecoveryFlags.ATTEMPT_RECOVERY,
```

Comment:
It is possible to choose if the provider will attempt to recover from an error by specifying flags when adding the provider to the simulation.